docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,184 @@
+| Code Fixer | A **code fixer** contains the code that fixes the violation by modifying source code. **Code fixes** in Visual Studio appear as "light bulb" suggestions. |
+| Language agnostic | We use this phrase to describe when the **analyzer** or **code fixer** is able to operate on both C# and Visual Basic.|
+
+<br />
+
+We use [Roslyn Analyzers](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview) to highlight areas of code that will need to be refactored. By default, we aim to pair these Analyzers with Code Fixers to automate as much of the upgrade workflow as possible.
+
+Example of changes that we can perform include:
+* Mapping from one type to another one common example is [System.Web.Http.ApiController](https://docs.microsoft.com/en-us/previous-versions/aspnet/hh834453(v=vs.118)) should become [Microsoft.AspNetCore.Mvc.ControllerBase](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase)
+* Extracting HttpContext via method dependency injection
+* Replacing methods that are not available on .NET latest such as [BinaryFormatter.UnsafeDeserialize](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter.unsafedeserialize)
+
+The following guidelines are a list of best practices that we use to guide the development of Analyzers and Code Fixers to support our goals.
+
+### Our Goals
+
+1. Write language agnostic analyzers with as much code reuse as possible.
+2. Analyzers should be as performant as possible.
+3. Samples included follow best practices as setup by [Roslyn Analyzers](https://github.com/dotnet/roslyn-analyzers).
+4. Use abstractions, and clean code principles, to write code that everyone can read regardless of their Roslyn experience to promote community contributions and reduce code maintenance.
+
+## Best Practices for Roslyn Analyzers and Code Fixes
+
+### 1. Use the *Microsoft.CodeAnalysis.Testing* framework
+
+Separation of analyzer and code fix tests increases complexity and code duplication and tends to decrease the overall confidence in the test suite. If you're testing entire files at a time, you will either be quickly overwhelmed by the number of files per test scenario or be tempted to put multiple test scenarios into a single file which shifts from unit to integration testing. The *Microsoft.CodeAnalysis.Testing* framework addresses these concerns.
+
+**Do**
+* Read the testing overview: [Microsoft.CodeAnalysis.Testing](https://github.com/dotnet/roslyn-sdk/blob/main/src/Microsoft.CodeAnalysis.Testing/README.md)
+
+### 2. Avoid member variables and state management
+Expect your analyzer to be invoked repeatedly and asynchronously. Design your analyzer so that execution can start processing a 2nd call before processing finishes for the 1st call. 
+
+**Do**
+* Pass information between methods via method arguments.
+
+**Do Not**
+* Do not use member variables to store instance data.
```

Comment:
this probably needs more clarity.  For example, you shoudl not store instance data on your analyzer or fixer instance itself.  However, both of those types can (and often should) instantiate a new type when they're called into that will store state and perform logic as appropriate.  This way you can use standard classes and wahtnot to organize things.  For example:

`context.RegisterSyntaxNodeAction(new SyntaxAnalyzer().Analyze, ...)`

SyntaxAnalyzer can now be a rich class that can do these things.